proposals/25-ipld-in-ipfs.md

```diff
@@ -0,0 +1,148 @@
+
+#### Assumptions &amp; hypotheses
+
+* As the foundational data model for our technology stack, the quality of our libraries and technologies for working with IPLD affect the entire web3 stack.
+* We’d like to have access to IPLD Selectors or other IPLD querying capabilities across the web3 stack for our own and others development efforts, because they enable a natural representation for addressing individual pieces of data.
+* There are external developers who would like to work directly with structured data in our stack, as opposed to serializing to flat files prior to import. Support for structured data in IPFS is immediately useful for them.
+* go-ipld-prime is sufficiently mature to be integrated through major portions of the web 3 stack, and remaining sharp edges are most easily ironed out through integration.
+
+#### User workflow example
+
+I am a developer.
+I have a set of structured JSON data.
+Or, perhaps I have an existing Web2 JSON API.
+I want to import that data directly into IPFS.
+
+I can run:
+
+    ipfs dag import users.json
+
+Or perhaps even:
+
+    ipfs dag import  https://api.spacexdata.com/v4/launches/latest
+
+This returns a CID for the data, encoded in a block in CBOR.
+I can fetch said block and output as JSON.
+
+Let’s say the CID for our SpaceX json is `QMABCDEF`.
+Soon, on another node, I should be able to fetch that data from the CID reference (note that selector pathing is not fully encompassed in this project):
+
+    ipfs dag get QMABCDEF/links/flickr/original/0
+    
+And per the sample response [here](https://github.com/r-spacex/SpaceX-API),
+I’d get back the text `https://live.staticflickr.com/65535/49927519643_b43c6d4c44_o.jpg`.
+
+Eventually, we might choose to recognize http links, download the results, and convert them to MerkleDAG data (either more CBOR data for json or UnixFS for HTML/Gif/PNG).
+Now the same query gets back `QMGHIJKL` -- a CID to download the image above of a space shuttle launch!
+
+#### Impact
+
+We have a number of existing users who have structured data they want to put in IPFS.
+Currently, they are expending effort to take that data and put it in flat files before importing into IPFS, with almost no query capabilities. 
+
+#### Leverage
+
+This a project whose greatest impact may be unblocking other projects. Among them:
+
+* Faster data transfer in IPFS (we can integrate bitswap and graphsync to speed up transfer, we cannot in the current stack)
+* Unlocking greater capabilities in the stack through use of selectors, including to-be-written selectors
+* IPLD based databases and graph querying as optimization mechanism
+* Standardizing IPLD across the stack including Filecoin (see meta doc above)
+* Coherent IPFS and Filecoin retrieval patterns (also needs Filecoin retrieval improvements)
+* Almost any effort to support dApps that work with non-static or non-file data
+
+#### Confidence
+
+On the scale: 1 -- Low
```

Comment:
Do we have any color as to why we have low confidence for this?

Reply:
I think you could say the tension goes something like this:

1. There's a *ton* of power in the IPLD stack as a full-functioning data system, as outlined here.
2. Developers have just got used to throwing all their data into raw blocks and wrapping structured data around DAG-PB / UnixFS and _just making it work_.

We've been pushing IPLD to do all sorts of powerful things ([e.g.](https://github.com/mikeal/ipsql)), and so have some of our more progressive [ecosystem partners](https://github.com/textileio/go-threads/tree/master/cbor). But perhaps these are just advanced users, and the average web3 dev is perfectly happy _treating everything as a file_, after all, that works perfectly well in unix!

Reply:
This was based on the doc that @anorth shared - https://medium.com/@nimay/inside-product-introduction-to-feature-priority-using-ice-impact-confidence-ease-and-gist-5180434e5b15 - which suggested internal assumptions, intuitions, and anecdotal evidence, did not amount to a high number. In general, we have very little data beyond this for almost any of the proposed projects.

Reply:
I think this was a conservative number based on the scale that we haven't done market research, which was indicated as a threshold for a 3 or above. we have heard things depending on this as a user ask from many existing users though, so arguably could push it harder.